nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilderFacetImpl.java

```diff
@@ -0,0 +1,198 @@
+import org.sonatype.nexus.repository.r.RHostedFacet;
+import org.sonatype.nexus.repository.r.RPackagesBuilderFacet;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.storage.AssetCreatedEvent;
+import org.sonatype.nexus.repository.storage.AssetDeletedEvent;
+import org.sonatype.nexus.repository.storage.AssetEvent;
+import org.sonatype.nexus.repository.storage.AssetUpdatedEvent;
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.transaction.UnitOfWork;
+
+import com.google.common.eventbus.AllowConcurrentEvents;
+import com.google.common.eventbus.Subscribe;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.State.STARTED;
+import static org.sonatype.nexus.repository.r.internal.AssetKind.ARCHIVE;
+import static org.sonatype.nexus.repository.r.internal.RPathUtils.getBasePath;
+
+/**
+ * Implementation of {@link RPackagesBuilderFacet} targeted for use with hosted repositories. Uses event dispatching to
+ * ensure that we do not have race conditions when processing and rebuilding metadata, and also imposes a waiting period
+ * to batch metadata updates.
+ */
+@Named
+public class RPackagesBuilderFacetImpl
+    extends FacetSupport
+    implements RPackagesBuilderFacet, Asynchronous
+{
+  /**
+   * The event manager to use when posting new events.
+   */
+  private final EventManager eventManager;
+
+  /**
+   * The interval in milliseconds to wait between rebuilds.
+   */
+  private final long interval;
+
+  /**
+   * The flag indicating if we are currently waiting to rebuild.
+   */
+  private boolean waiting;
+
+  /**
+   * Constructor.
+   *
+   * @param eventManager The event manager to use when posting new events.
+   * @param interval     The interval in milliseconds to wait between rebuilds.
+   */
+  @Inject
+  public RPackagesBuilderFacetImpl(
+      final EventManager eventManager,
+      @Named("${nexus.r.packagesBuilder.interval:-1000}") final long interval)
+  {
+    this.eventManager = checkNotNull(eventManager);
+    this.interval = interval;
+  }
+
+  /**
+   * Handles {@code AssetDeletedEvent} events concurrently, requesting a metadata invalidation and rebuild if warranted
+   * by the event contents.
+   *
+   * @param deleted The deletion event to handle.
+   */
+  @Subscribe
+  @Guarded(by = STARTED)
+  @AllowConcurrentEvents
+  public void on(final AssetDeletedEvent deleted) {
+    if (shouldInvalidate(deleted)) {
+      invalidateMetadata(getBasePath(deleted.getAsset().name()));
+    }
+  }
+
+  /**
+   * Handles {@code AssetCreatedEvent} events concurrently, requesting a metadata invalidation and rebuild if warranted
+   * by the event contents.
+   *
+   * @param created The creation event to handle.
+   */
+  @Subscribe
+  @Guarded(by = STARTED)
+  @AllowConcurrentEvents
+  public void on(final AssetCreatedEvent created) {
```

Comment:
Question: Could these 3 methods will simplify to use general AssetEvent (there are only 3 current events) and use only one method instead?

Reply:
It may fail if any more asset events are added

Reply:
But you can check directly in one method and do not create 3

Reply:
Discussed and decided to put into single `on()` method. Done in f0536c518860f9b4d4eadd343d2585cb39f60eb3